### PR TITLE
[RBS-952] validation and schema for nested-multi-select

### PR DESCRIPTION
**Pull request by Mono143**
**Changes**
Added new validation and schema for nested-multi-select
Fixed critical vulnerability

-   [delete] branch

**Additional information**

-   You may refer to [RBS-952](https://jira.ship.gov.sg/browse/RBS-952)

### CODE REVIEW

**Review by qroll — DISMISSED** [review]
thanks for submitting the PR and updating the deps

backend validation has to be stricter as there's no client to help with filtering of options

could include additional `test`s that:
1. selected items indeed belong to the options
2. at least 1 item is selected e.g. `{ fruitKey: {} }` should fail as the child item was not actually selected

File: src/fields/nested-multi-select.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import * as Yup from "yup";
+import { IFieldSchemaBase } from "../schema-generator";
+import { ERROR_MESSAGES } from "../shared";
+import { IFieldGenerator } from "./types";
+import { has } from "lodash";
```
Comment (weili-govtech) [inline src/fields/nested-multi-select.ts:5]: remove unused import

File: src/fields/nested-multi-select.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	subItems?: TL3OptionProps[] | undefined;
+}
+export interface TL3OptionProps extends BaseOptionProps {
+	value: string;
+	subItems?: undefined;
+}
+
+export interface INestedMultiSelectSchema<V = undefined> extends IFieldSchemaBase<"nested-multi-select", V> {
+	options: TL1OptionProps[];
+}
```
Comment (weili-govtech) [inline src/fields/nested-multi-select.ts:36]: - there are many typings declared, can move them to a `types.ts` file for better readability
- interfaces and types should be prepended by `I` and `T` respectively
- for L1 and L2 option props, they can come with either a `value` or `subItems` but never both, current typing allows both, you can prevent this via the union type, e.g.

```
interface IBaseOptionPropsWithValues extends BaseOptionProps {
	value: string;
	subItems?: never;
}

interface IBaseOptionPropsWithSubItems<T> extends BaseOptionProps {
	value?: never;
	subItems: T | undefined;
}

export type TL1OptionProps = IBaseOptionPropsWithValues | IBaseOptionPropsWithSubItems<TL2OptionProps[]>;
```

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: src/fields/nested-multi-select.ts (hunk 2 above)
Comment (Mono143) [reply src/fields/nested-multi-select.ts]: thanks for the comments, updated the pr based on your suggestions. 
the typings were probably remnants of my previous implementation, sorry bout that 😓

File: src/fields/nested-multi-select/types.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { IFieldSchemaBase } from "../../schema-generator";
+
+export interface BaseOptionProps {
```
Comment (weili-govtech) [inline src/fields/nested-multi-select/types.ts:3]: `IBaseOptionProps`

File: src/fields/nested-multi-select/types.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	key: string;
+}
+interface IBaseOptionPropsWithValues extends BaseOptionProps {
+	value: string;
+	subItems?: never;
+}
+
+interface IBaseOptionPropsWithSubItems<T> extends BaseOptionProps {
+	value?: never;
+	subItems: T | undefined;
```
Comment (weili-govtech) [inline src/fields/nested-multi-select/types.ts:14]: whoops sorry, this should not allow undefined

**Review by weili-govtech — CHANGES_REQUESTED** [review]

**Review by weili-govtech — APPROVED** [review]

**Pull request merged**